library/init/meta/interactive.lean

```diff
@@ -375,14 +375,55 @@ do e_type ← infer_type e >>= whnf,
    (const I ls) ← return $ get_app_fn e_type,
    return I
 
+meta def cases_arg_p : parser (option name × pexpr) :=
+with_desc "(id :)? expr" $ do
+  t ← texpr,
+  match t with
+  | (local_const x _ _ _) :=
+    (tk ":" *> do t ← texpr, pure (some x, t)) <|> pure (none, t)
+  | _ := pure (none, t)
+  end
+
+private meta def generalize_arg_p : parser (pexpr × name) :=
+with_desc "expr = id" $ do
+  e ← parser.pexpr 0,
+  match e with
+  | (app (app (macro _ [const `eq _ ]) h) (local_const x _ _ _)) := pure (h, x)
+  | _ := fail "parse error"
+  end
+
+/-- `generalize : e = x` replaces all occurrences of `e` in the target with a new hypothesis `x` of the same type.
+    `generalize h : e = x` in addition registers the hypothesis `h : e = x`. -/
+meta def generalize (h : parse ident?) (_ : parse $ tk ":") (p : parse generalize_arg_p) : tactic unit :=
```

Comment:
@leodemoura Is this better? I'm torn on whether to parse `:` in a separate arg like this or use `*>` like before; this way is a bit clearer but also requires an extra dummy arg when called directly a few times below.

Reply:
Yes, it is better. It would be even better if you add comments to `generalize_arg_p` and other functions you have added.

Reply:
I added some provisional descriptions of `by_cases` and `induction`, which were modified in this PR, but I will defer to @avigad who is also working on inline documentation of the tactics here.